tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/mailet/CalDavCollect.java

```diff
@@ -0,0 +1,132 @@
+import com.linagora.tmail.dav.EventUid;
+
+import net.fortuna.ical4j.model.Calendar;
+import net.fortuna.ical4j.model.Property;
+import net.fortuna.ical4j.model.component.VEvent;
+import net.fortuna.ical4j.model.property.Method;
+import net.fortuna.ical4j.model.property.Uid;
+import reactor.core.publisher.Mono;
+
+public class CalDavCollect extends GenericMailet {
+    public static final String SOURCE_ATTRIBUTE_NAME = "source";
+    public static final String DEFAULT_SOURCE_ATTRIBUTE_NAME = "calendars";
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CalDavCollect.class);
+    private static final Class<Map<String, AttributeValue<Calendar>>> MAP_STRING_CALENDAR_CLASS = (Class<Map<String, AttributeValue<Calendar>>>) (Object) Map.class;
+
+    private final DavClient davClient;
+    private final DavUserProvider davUserProvider;
+
+    private AttributeName sourceAttributeName;
+
+    @Inject
+    public CalDavCollect(DavClient davClient, DavUserProvider davUserProvider) {
+        this.davClient = davClient;
+        this.davUserProvider = davUserProvider;
+    }
+
+    @Override
+    public void init() throws MessagingException {
+        sourceAttributeName = AttributeName.of(getInitParameter(SOURCE_ATTRIBUTE_NAME, DEFAULT_SOURCE_ATTRIBUTE_NAME));
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        try {
+            AttributeUtils.getValueAndCastFromMail(mail, sourceAttributeName, MAP_STRING_CALENDAR_CLASS)
+                .ifPresent(calendars -> calendars.values()
+                    .stream()
+                    .map(AttributeValue::getValue)
+                    .toList()
+                    .forEach(calendar -> handleCalendarInMail(calendar, mail)));
+        } catch (ClassCastException e) {
+            LOGGER.error("Received a mail with {} not being an ICAL object for mail {}", sourceAttributeName, mail.getName(), e);
+        }
+    }
+
+    private void handleCalendarInMail(Calendar calendar, Mail mail) {
+        try {
+            mail.getRecipients()
+                .forEach(mailAddress -> davUserProvider.provide(Username.of(mailAddress.asString()))
+                    .flatMap(davUser -> {
+                        EventUid eventUid = getEventUid(calendar);
+                        if (Method.VALUE_CANCEL.equals(calendar.getMethod().getValue())) {
+                            return deleteDavCalendar(davUser, eventUid);
+                        } else {
+                            return davClient.getCalendarObject(davUser, eventUid)
+                                .flatMap(davCalendarObject -> updateDavCalendar(davCalendarObject, calendar, davUser))
+                                .switchIfEmpty(createDavCalendar(calendar, davUser));
+                        }
+                    }).block());
+        } catch (Exception e) {
+            LOGGER.error("Error while handling calendar in mail {}", mail.getName(), e);
+        }
+    }
```

Comment:
Error handling sucks

If I send an email with ICS to bob and alice and that somehow we fail to update bob calendar, there is no reasons not to update Alice calendar

Please adapt exception handling accordingly.